File: C++/022. Generate Parentheses.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    std::vector<std::string> generateParenthesis(int n) {
        std::vector<std::string> res;
        helper(res, "", n, 0);
        return res;
    }

    void helper(std::vector<std::string> &res, std::string str, int left_par, int right_par) {
        if (left_par == 0 && right_par == 0) {
            res.push_back(str);
            return;
        }
        if (right_par > 0)
            helper(res, str + ")", left_par, right_par - 1);
        if (left_par > 0)
            helper(res, str + "(", left_par - 1, right_par + 1);
    }
};

File: Go/022. Generate Parentheses.go
package main

func generateParenthesis(n int) []string {
	res := make([]string, 0, 0)
	return helper(res, "", n, 0)
}

func helper(res []string, str string, left int, right int) []string{
	if left == 0 && right == 0 {
		res = append(res, str)
		return res
	}
	if right > 0 {
		res = helper(res, str + ")", left, right - 1)
	}
	if left > 0 {
		res = helper(res, str + "(", left - 1, right + 1)
	}
	return res
}